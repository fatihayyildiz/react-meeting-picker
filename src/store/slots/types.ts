// Type declarations of models
export type Company = {
	id: number
	name: string
	type: string
	time_slots: Array<TimeSlot>
}

export type TimeSlot = {
	start_time: Date
	end_time: Date
}

export type SlotsInitialState = {
	loading: boolean
	companies: Array<Company>
}