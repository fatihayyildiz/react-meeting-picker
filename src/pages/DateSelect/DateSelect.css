.date-select-container > .row {
	overflow-x: scroll;
	display: block;
	white-space: nowrap;
}

.date-select-container > .row > .col-3 {
	display: inline-block;
	min-height: 200px;
	vertical-align: top;
}

.centered-box {
	padding: 10px 20px;
	text-align: center;
	border: 1px solid #e1e1e1;
	border-radius: 2px;
	margin-top: 20px !important;
	background-color: white;
}

.fit-height-box {
	max-height: 100vh;
	overflow-y: auto;
}

.extra-info-container {
	margin-top: 10px !important;
}
