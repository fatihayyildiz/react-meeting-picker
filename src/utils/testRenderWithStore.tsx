import React from 'react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { render as renderRtl } from '@testing-library/react';
import { RootState } from 'store/store';
import { reducers } from 'store/reducers';

interface WrapperProps {
	children?: React.ReactNode;
}

const testStore = (state: Partial<RootState>) => {
	return configureStore({
		reducer: reducers,
		preloadedState: state,
	});
};

export const testRenderWithStore = (
	component: React.ReactNode,
	initialState: any
) => {
	const wrapper = ({ children }: WrapperProps) => (
		<Provider store={testStore(initialState)}>{children}</Provider>
	);
	return renderRtl(component as any, { wrapper });
};
