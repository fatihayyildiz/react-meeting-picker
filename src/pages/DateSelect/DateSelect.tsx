import React from 'react';
import 'pages/DateSelect/DateSelect.css';
import CompanyContainer from 'containers/CompanyContainer/CompanyContainer';
import FormattedCompanies from 'mocks/resources/formatted_companies.json';

import { DaySlotsProps } from 'containers/DaySlots/DaySlots';

const DateSelect = () => {
	return (
		<div className="container date-select-container">
			<div className="row">
				{FormattedCompanies.map((company, index) => {
					return (
						<div className="col-3 fit-height-box" key={`company-${index}`}>
							<CompanyContainer
								key={`company-container-${index}`}
								name={company.name}
								days={company.days as unknown as Array<DaySlotsProps>}
							/>
						</div>
					);
				})}
			</div>
		</div>
	);
};

export default DateSelect;
