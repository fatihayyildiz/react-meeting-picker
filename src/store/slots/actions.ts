import { createAsyncThunk } from '@reduxjs/toolkit';
import { SlotsApi } from 'store/slots/api';

const fetchSlots = createAsyncThunk(
	'slots/fetchSlots',
	async () => {
		try{
			// Call axios instance to get slots
			return await SlotsApi.getInstance().getAllSlots()
		}catch(error: unknown){
			console.error('fetch slots api call error:',error)
			throw Error('fetch slots api call error')
		}
	});

export {fetchSlots}