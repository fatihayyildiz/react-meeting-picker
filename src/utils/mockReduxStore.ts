import { CompanySelectedSlot, CompanySlots } from 'store/slots/types';

export const MockReduxStoreInitialState = {
	slots: {
		loading: false,
		companies: [] as Array<CompanySlots>,
		selectedCompanySlots: [] as Array<CompanySelectedSlot>,
	},
};
