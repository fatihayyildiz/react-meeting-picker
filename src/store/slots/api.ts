import { HttpClient } from 'utils/HttpClient';
import { APICompany } from 'store/slots/types';

export class SlotsApi extends HttpClient {
	private static classInstance?: SlotsApi;

	constructor() {
		super('http://localhost:3000/api');
	}

	// To avoid create an instance everytime, pointed to Singleton of static value
	public static getInstance() {
		if (!this.classInstance) {
			this.classInstance = new SlotsApi();
		}
		return this.classInstance;
	}

	public getAllSlots = () => this.instance.get<Array<APICompany>>(`/slots`);
}
