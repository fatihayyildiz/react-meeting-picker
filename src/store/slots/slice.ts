import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { APICompany, CompanySelectedSlot, CompanySlots, SlotsInitialState } from 'store/slots/types';
import { fetchSlots } from 'store/slots/actions';
import { RootState } from 'store/store';
import { convertApiResponseToFEModel } from 'utils/dataMapper';
import { TimeSlotProps } from 'components/TimeSlot/TimeSlot';

const initialState: SlotsInitialState = {
	loading: false,
	companies: [] as Array<CompanySlots>,
	selectedCompanySlots: [] as Array<CompanySelectedSlot>,
};

const slotSlice = createSlice({
	name: 'slots',
	initialState,
	reducers: {
		selectSlotForCompany: (state, action: PayloadAction<TimeSlotProps>) => {
			// Check for selected slot exist
			const checkSelectedSlotIndex = state.selectedCompanySlots.findIndex((slot) => slot.id === action.payload.id);
			// If agent select slot before, change that slot with new selected one
			if (checkSelectedSlotIndex > -1)
				state.selectedCompanySlots[checkSelectedSlotIndex] = {
					id: action.payload.id,
					selectedSlot: {
						day: action.payload.day,
						start_time: action.payload.start_time,
						end_time: action.payload.end_time
					},
				};
				// Agent didnt selected any slot for this company
			// push new slot to store
			else
				state.selectedCompanySlots.push({
					id: action.payload.id,
					selectedSlot: {
						day: action.payload.day,
						start_time: action.payload.start_time,
						end_time: action.payload.end_time
					},
				});
		},
		removeSlotForCompany: (state,action:PayloadAction<TimeSlotProps>)=>{
			const checkSelectedSlotIndex = state.selectedCompanySlots.findIndex((slot) => slot.id === action.payload.id);
			console.log('check slot index:',checkSelectedSlotIndex,action.payload)
			//const shallowCopy = JSON.parse(JSON.stringify(state.selectedCompanySlots))
			state.selectedCompanySlots.splice(checkSelectedSlotIndex,checkSelectedSlotIndex+1)
			//console.log('state.selectedCompanySlots:',shallowCopy)
			//state.selectedCompanySlots = shallowCopy
			//console.log('new company slots:',newCompanySlots)
			//state.selectedCompanySlots = newCompanySlots

		}
	},
	extraReducers: (builder) => {
		builder.addCase(fetchSlots.pending, (state) => {
			state.loading = true;
			state.companies = [] as Array<CompanySlots>;
			state.selectedCompanySlots = [] as Array<CompanySelectedSlot>;
		});

		builder.addCase(fetchSlots.fulfilled, (state, { payload }) => {
			state.loading = false;
			const formattedCompanySlots = convertApiResponseToFEModel(payload as Array<APICompany>);
			state.companies = formattedCompanySlots as Array<CompanySlots>;
			state.selectedCompanySlots = [] as Array<CompanySelectedSlot>;
		});

		builder.addCase(fetchSlots.rejected, (state) => {
			state.loading = false;
			state.companies = [] as Array<CompanySlots>;
			state.selectedCompanySlots = [] as Array<CompanySelectedSlot>;
		});
	},
});

export const slotSelector = (state: RootState): SlotsInitialState => state.slots;

export default slotSlice;