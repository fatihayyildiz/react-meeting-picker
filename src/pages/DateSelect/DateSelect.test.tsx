import React from 'react';
import {
	cleanup,
	screen,
	RenderResult,
	waitFor,
	fireEvent,
} from '@testing-library/react';
import DateSelect from 'pages/DateSelect/DateSelect';
import { testRenderWithStore } from 'utils/testRenderWithStore';
import { MockReduxStoreInitialState } from 'utils/mockReduxStore';
import MockTimeSlots from 'mocks/resources/time_slots.json';
import { APICompany } from 'store/slots/types';
import { server } from 'mocks/server.js';

const renderDateSelectPage = (): RenderResult =>
	testRenderWithStore(<DateSelect />, MockReduxStoreInitialState);

afterEach(cleanup);

// Establish API mocking before all tests.
beforeAll(() => server.listen());

// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers());

// Clean up after the tests are finished.
afterAll(() => server.close());

test('Date Select page should render loading info', async () => {
	// First render component and get as fragment
	const { asFragment } = renderDateSelectPage();

	// Compare with saved snapshot
	expect(asFragment()).toMatchSnapshot();
});

test('Date Select page should render company name and slots correctly', async () => {
	// First render component and get as fragment
	renderDateSelectPage();

	// wait for page load api data
	await waitFor(() => {
		// Get first company from mock dataset
		const firstCompany = MockTimeSlots[0] as unknown as APICompany;
		expect(screen.getByText(firstCompany.name)).toBeInTheDocument();
	});

	await waitFor(() => {
		const apiCompanyList = MockTimeSlots as unknown as Array<APICompany>;
		const timeSlotsCount = apiCompanyList.reduce((acc, curr) => {
			acc = acc + curr.time_slots.length;
			return acc;
		}, 0);
		const screenTimeSlots = screen.getAllByTestId('time-slot');
		expect(screenTimeSlots.length).toBe(timeSlotsCount);
	});
});

test('Time slot click action should add selected class', async () => {
	// First render component and get as fragment
	const { container } = renderDateSelectPage();

	// Wait until time slots elements appear at screen
	await screen.findAllByTestId('time-slot');

	// get all timeslots at screen
	const screenTimeSlots = screen.getAllByTestId('time-slot');

	// Click first time slot object
	fireEvent.click(screenTimeSlots[0]);

	// eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
	const selected = container.getElementsByClassName('selected');

	expect(selected.length).toBe(1);
});

test('Reservation clear button should remove selected class', async () => {
	// First render component and get as fragment
	const { container } = renderDateSelectPage();

	// Wait until time slots elements appear at screen
	await screen.findAllByTestId('time-slot');

	// get all timeslots at screen
	const screenTimeSlots = screen.getAllByTestId('time-slot');

	// Click first time slot object
	fireEvent.click(screenTimeSlots[0]);

	// eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
	const selected = container.getElementsByClassName('selected');

	// Wait until reservation clear elements appear at screen
	await screen.findAllByTestId('clear');

	// eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
	const clearButton = screen.getAllByTestId('clear');

	// Selected timeslot should appear at screen
	expect(selected.length).toBe(1);

	// Clear reservation button should appear at screen
	expect(clearButton.length).toBe(1);

	// Click clear reservation button
	fireEvent.click(clearButton[0]);

	// Check clear button disappear from screen
	await waitFor(() => {
		expect(clearButton[0]).not.toBeInTheDocument();
	});
});
