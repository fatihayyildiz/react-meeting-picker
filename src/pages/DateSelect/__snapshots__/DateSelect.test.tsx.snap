// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date Select page should render loading info 1`] = `
<DocumentFragment>
  <div
    class="container date-select-container"
  >
    <div
      class="row"
    >
      <span>
        Loading ...
      </span>
    </div>
  </div>
</DocumentFragment>
`;
